client/web/src/cody/management/api/hooks/useApiClient.tsx

```diff
@@ -0,0 +1,85 @@
+import { useEffect, useState, useContext } from 'react'
+
+import { Call } from '../client'
+import { CodyProApiClientContext } from '../components/CodyProApiClient'
+
+export interface ReactFriendlyApiResponse<T> {
+    loading: boolean
+    error?: Error
+    data?: T
+    response?: Response
+}
+
+// useApiCaller will issue a REST API call to the backend, returning the
+// loading status and the response JSON object and/or error object as React
+// state.
+//
+// IMPORTANT: In order to avoid the same API request being made multiple times,
+// you MUST ensure that the provided call is the same between repains of the
+// calling React component. i.e. you pretty much always need to create it via
+// `useMemo()`.
+export function useApiCaller<Resp>(call: Call<Resp>): ReactFriendlyApiResponse<Resp> {
+    const { caller } = useContext(CodyProApiClientContext)
+
+    const [loading, setLoading] = useState(true)
+    const [error, setError] = useState<any>(undefined)
+    const [data, setData] = useState<any>(undefined)
+    const [response, setResponse] = useState<any>(undefined)
```

Comment:
You can definitely type these more strongly:

```suggestion
    const [error, setError] = useState<Error | undefined>(undefined)
    const [data, setData] = useState<Resp | undefined>(undefined)
    const [response, setResponse] = useState<Response | undefined>(undefined)
```

Reply:
Done.